### ISSUE TITLE
Plot insets (feature request)

### ISSUE THREAD

**Issue by basnijholt** [ioam/holoviews#175]
I would like to put small plots inside my Images, like is done in: http://matplotlib.org/1.3.1/mpl_toolkits/axes_grid/users/overview.html

**Comment by jlstevens** [comment]
From the discussion on Gitter, here is what Philipp suggested:

```python
def inset_hook(plot, view):
    ax = plot.handles['axis']
    fig = plot.handles['fig']
    ...
```

Then you set the plot option ``finalize_hooks=[inset_hook]`` and can do whatever you like with the axis and figure object inside the function

Philipp also said he might have a look at supporting this feature more explicitly soon.

**Comment by philippjfr** [comment]
The main issue is deciding how to set the inset options. First of all there's two types of inset:

```python
inset_axes(parent_axes, width="30%", # width = 30% of parent_bbox
                  height=1., # height : 1 inch
                  loc=3)
```

and 

```python
inset_axes = zoomed_inset_axes(ax, 0.5, # zoom = 0.5
                                                      loc=1)
```

The signatures are similar but not the same, so the only thing I can think of is having a inset_options dictionary as a plot options, which if specified applies the inset. Thoughts?

**Comment by jlstevens** [comment]
Hmm, I'm not sure how I feel about insets being set via a plot option. 

To me the concept of an inset is another kind of composition and something that we might want to support in a general way. I know it is a useful feature to support but I am not yet sold on this particular interface.

**Comment by jbednar** [comment]
I agree -- insets have very wide applicability, and it is important to really understand how they are used in the wild so that very general support can be provided. Working up a quick demo is great,, but being sure it will really solve the issue will probably take a while. Seems very related to the issue of AdjointLayout.

**Comment by philippjfr** [comment]
Another operator seems like overkill when overlaying already does most of the work, in fact you can already inset something by setting the desired extents. The only thing we'd be adding is support for drawing an additional set of axes inside an existing set of axes. Adding new operators/containers to do display level stuff is a recipe for nightmares if AdjointLayout is anything to go by. It's also unclear in how far different plotting backends will support the idea of insets so adding specific operators or objects to support them may not be very general. IMO insetting should definitely happen by overlaying the inset and setting the plot options, I'm almost certain adding new operators/classes for this is a bad idea.

**Comment by jlstevens** [comment]
>  I'm almost certain adding new operators/classes for this is a bad idea

We haven't discussed anything at all yet. I'm just saying that plot options are a pretty poor interface for such a common type of visualization.

**Comment by philippjfr** [comment]
That's why we're discussing it now. There's only a few interfaces we have to support this kind of thing, so it's either a new kind of operator/class, a plot option or some state on the objects itself (which is what we're trying to avoid). All I'm saying is that out of those options, setting it via a plot option is by far the most palatable to me.

**Comment by jlstevens** [comment]
You are only thinking of a few limited possibilities - plot options, new class, new operators. None  of these approaches are required to come up with an approach superior to using plot options.

Here is a proposal that makes much more sense to me:

* Insets are indeed related to ``Overlays`` and in fact that seems the key: overlays should be concerned with insets.
* Insets are in fact a special case of what overlays already achieve: an inset the size of the original plot is exactly the way ``Overlay`` objects already behave.
* The position of an inset is part of the data if you consider generalizing the concept of an insets. Let's say, I have some contours in a figure but I only want to overlay these contours in the top-left (i.e just like an inset). Then because you can see the data underneath the contour lines, the position of the inset in relation to the data matters. Even if you disregard this point, the inset position is not a backend dependent setting.

In conclusion,I think building an inset should be as simple as calling a method on ``Overlay`` to specify the portion of space occupied by the inset layers. Something like this is definitely worth considering before delegating such an crucial feature to the plot options.

**Comment by philippjfr** [comment]
> The position of an inset is part of the data if you consider generalizing the concept of an insets. Let's say, I have some contours in a figure but I only want to overlay these contours in the top-left (i.e just like an inset). Then because you can see the data underneath the contour lines, the position of the inset in relation to the data matters (i.e the data between layers needs may need to be aligned and in the same space). Even if you disregard this point, the inset position is not a backend dependent setting.

Not sure I quite follow but whatever we end up doing the Overlay will have to somehow group inset objects together, probably on the inset bounds. 

>In conclusion, I think building an inset should be as simple as calling a method on Overlay to specify the portion of space occupied by the inset layers.

To me this is still less clear and flexible than controlling it via plot options. There's different kinds of insets in matplotlib, something that a general interface like this could never account for. Secondly it breaks compositionality, you can't declare a single Element as an inset and then reuse that on multiple plots. You'd either have to do this each time you want to create an inset:

```python
(img1 * img2).inset(index=1, **inset_kwargs)
```
which declares the second Image as the inset via an index or it's path name or more explicitly do this:
```python
Overlay([img]).inset(img2, **inset_kwargs)
```
which seems even more awkward to me. I therefore still think doing it via the plot options is the clearest and cleanest solution, as it would look something like this:
```python
img * img2(plot={'inset': inset_kwargs})
```
Sounds like we'll have to discuss it more in person to come to a conclusion though.

**Comment by jlstevens** [comment]
I'll just say that:

```python
(img1 * img2).inset(index=1, **inset_kwargs)
```

Seems no worse to me than:

```python
img * img2(plot={'inset': inset_kwargs})
```

I'm not sure what I think of having inset defined at the element level and not the overlay level. Your example here works because of ``__call__`` which you shouldn't have to use. The idea of insets being a plot option in the ``%opts`` magic doesn't particularly appeal to me.

I'll have to think about it.

**Comment by philippjfr** [comment]
I might be coming around to the method for another reason. My proposal would require Overlays to be nestable to become easily usable when insetting an Overlay. Say you have two Overlays and want to inset one in the other, when using the method you could theoretically do

```python
overlay1 * overlay2.index(index=range(len(overlay2)), **inset_kwargs)
```

Since the Overlay knows about insets, it can retain the inset settings when performing the ``__mul__``, while in my proposal the inset options would get lost when doing:

```python
overlay1 * overlay2(plot={'inset': inset_kwargs})
```

unless the ``__mul__`` operations knows to nest Overlays, which means it would have to inspect the plot options. The main reason why I'm uncomfortable about a method handling this is because it would inevitably also have to store some backend specific options to be able to take advantage of all the features that matplotlib insets support. However I think we could consider a hybrid approach where any backend specific options are declared as plot options automatically (we already do something similar for the ``hist`` method).

**Comment by jlstevens** [comment]
> ... it would inevitably also have to store some backend specific options to be able to take advantage of all the features that matplotlib insets support.

Right, this is something to discuss. Earlier, you also said:

> There's different kinds of insets in matplotlib, something that a general interface like this could never account for.

The thing I think we need to figure out is *what* is the *semantic* differences between the matplotlib approaches?

I think HoloViews should come up with a general model of how insets work regardless of matplotlib or any particular backend. It needs to be simple and general so that you might expect any backend to be able to support it. What are the *real* differences in the various ways matplotlib supports insets that we should consider and maybe think of generalizing?

**Comment by philippjfr** [comment]
> The thing I think we need to figure out is what is the semantic differences between the matplotlib approaches?

matplotlib has two types of insets. Completely independent inset axes and zoomed inset axes (i.e. where the inset axes are in the same coordinate system as the host axis). 

>I think HoloViews should come up with a general model of how insets work regardless of matplotlib or any particular backend. It needs to be simple and general so that you might expect any backend to be able to support it.

I agree if we go with the method approach we need to have at least a minimal interface. We'll actually have to investigate the matplotlib implementation to see if it can be easily generalized. Both inset functions have a loc parameter, which accepts standard matplotlib locations as integers or strings. I'd hope that was an orthogonal option to supplying an explicit bounds but knowing matplotlib I'm not so sure they've provided a general interface that can be sensibly wrapped.

**Comment by jbednar** [comment]
Hmm. Don't Overlays all share the same axes?  If so the types of inset you discuss above don't seem like overlays, though they and Overlays could both be an example of some more general concept of being on  top of each other.  

I'm not sure what all Matplotlib does with insets, or what other packages do...

**Comment by philippjfr** [comment]
I definitely would not want to introduce another class, overlaying is still overlaying even if part of the overlay is in a new axis. Maybe it is a good idea after all for Overlays to be nestable because that way everything inside one axis is always kept together at the same level, while nested axes are represented by nested Overlays.
